File: ml/vproject_ml/features.py
from __future__ import annotations

from datetime import date, datetime, timedelta, timezone
from typing import Any

LEVELS = [
    (1, 0),
    (2, 200),
    (3, 600),
    (4, 1400),
    (5, 3000),
    (6, 6000),
    (7, 10000),
    (8, 16000),
    (9, 25000),
    (10, 40000),
    (11, 65000),
    (12, 100000),
]

FEATURE_COLUMNS = [
    "dias_ativos_7",
    "dias_ativos_21",
    "dias_sem_habito",
    "media_habitos_dia_7",
    "media_habitos_dia_21",
    "taxa_conclusao_7",
    "taxa_conclusao_21",
    "wd_0",
    "wd_1",
    "wd_2",
    "wd_3",
    "wd_4",
    "wd_5",
    "wd_6",
    "streak_atual",
    "streak_maximo",
    "xp_total",
    "nivel",
    "desafios_ativos",
    "desafios_concluidos_21",
    "desafios_expirados_21",
    "dias_desde_ultima_atividade",
    "media_xp_dia_21",
    "tomorrow_weekday",
    "tomorrow_rate",
]


def parse_day(iso: str) -> date:
    return date.fromisoformat(iso[:10])


def add_days(iso: str, n: int) -> str:
    return (parse_day(iso) + timedelta(days=n)).isoformat()


def day_diff(a_iso: str, b_iso: str) -> int:
    return (parse_day(b_iso) - parse_day(a_iso)).days


def calcular_nivel(xp: int) -> dict[str, Any]:
    atual = LEVELS[0]
    proximo = LEVELS[1] if len(LEVELS) > 1 else None
    for i, level in enumerate(LEVELS):
        if xp >= level[1]:
            atual = level
            proximo = LEVELS[i + 1] if i + 1 < len(LEVELS) else None
    xp_para = max(0, proximo[1] - xp) if proximo else 0
    return {
        "nivel": atual[0],
        "proximo": proximo[0] if proximo else None,
        "xp_para_proximo": xp_para,
    }


def compute_user_features(
    *,
    as_of_date: str,
    habit_count_ativo: int,
    completions: list[dict[str, Any]],
    challenges: list[dict[str, Any]],
    streak_atual: int,
    streak_maximo: int,
    xp_total: int,
    ultimo_dia_completo: str | None,
) -> dict[str, Any]:
    """Espelho de src/lib/ml/features.ts computeUserFeatures."""
    habit_n = max(0, habit_count_ativo)
    from7 = add_days(as_of_date, -6)
    from21 = add_days(as_of_date, -20)

    by_day: dict[str, int] = {}
    xp_by_day: dict[str, float] = {}
    for c in completions:
        dia = c["dia"][:10]
        if dia < from21 or dia > as_of_date:
            continue
        by_day[dia] = by_day.get(dia, 0) + 1
        xp = c.get("xp_ganho") or 0
        xp_by_day[dia] = xp_by_day.get(dia, 0) + float(xp)

    dias_ativos_7 = dias_ativos_21 = 0
    sum_habits_7 = sum_habits_21 = 0.0
    sum_xp_21 = 0.0
    dias_completos_7 = dias_completos_21 = 0

    for i in range(21):
        dia = add_days(from21, i)
        count = by_day.get(dia, 0)
        in7 = dia >= from7
        if count > 0:
            dias_ativos_21 += 1
            if in7:
                dias_ativos_7 += 1
        sum_habits_21 += count
        sum_xp_21 += xp_by_day.get(dia, 0)
        if in7:
            sum_habits_7 += count
        if habit_n > 0 and count >= habit_n:
            dias_completos_21 += 1
            if in7:
                dias_completos_7 += 1

    weekday_done = [0] * 7
    weekday_total = [0] * 7
    for i in range(21):
        dia = add_days(from21, i)
        wd = parse_day(dia).weekday()  # Mon=0 .. Sun=6 in Python
        # Align with JS getUTCDay: Sun=0 .. Sat=6
        wd_js = (wd + 1) % 7
        weekday_total[wd_js] += 1
        count = by_day.get(dia, 0)
        ok = count >= max(1, (habit_n + 1) // 2) if habit_n > 0 else count > 0
        if ok:
            weekday_done[wd_js] += 1

    weekday_rates = {
        str(wd): round(weekday_done[wd] / (weekday_total[wd] or 1), 4) for wd in range(7)
    }

    last_activity = None
    for dia in sorted(by_day.keys()):
        if by_day.get(dia, 0) > 0:
            last_activity = dia
    if ultimo_dia_completo:
        ud = ultimo_dia_completo[:10]
        if not last_activity or ud > last_activity:
            last_activity = ud

    dias_desde = day_diff(last_activity, as_of_date) if last_activity else None
    if dias_desde is not None:
        dias_desde = max(0, dias_desde)

    dias_sem_habito = 0
    for i in range(21):
        dia = add_days(as_of_date, -i)
        if by_day.get(dia, 0) == 0:
            dias_sem_habito += 1
        else:
            break

    from21_chal = f"{from21}T00:00:00.000Z"
    desafios_ativos = desafios_concluidos_21 = desafios_expirados_21 = 0
    for ch in challenges:
        status = ch.get("status")
        if status == "ativo":
            desafios_ativos += 1
        if status == "concluido":
            completed = ch.get("completed_at")
            if completed and completed >= from21_chal:
                desafios_concluidos_21 += 1
        if status == "expirado":
            ref = ch.get("ends_at") or ch.get("created_at")
            if ref and ref >= from21_chal:
                desafios_expirados_21 += 1

    level = calcular_nivel(xp_total)

    return {
        "features_version": "v1",
        "dias_ativos_7": dias_ativos_7,
        "dias_ativos_21": dias_ativos_21,
        "dias_sem_habito": dias_sem_habito,
        "media_habitos_dia_7": round(sum_habits_7 / 7, 3),
        "media_habitos_dia_21": round(sum_habits_21 / 21, 3),
        "taxa_conclusao_7": round(
            (dias_completos_7 / 7) if habit_n > 0 else (dias_ativos_7 / 7), 4
        ),
        "taxa_conclusao_21": round(
            (dias_completos_21 / 21) if habit_n > 0 else (dias_ativos_21 / 21), 4
        ),
        "weekday_rates": weekday_rates,
        "streak_atual": streak_atual,
        "streak_maximo": streak_maximo,
        "xp_total": xp_total,
        "nivel": level["nivel"],
        "desafios_ativos": desafios_ativos,
        "desafios_concluidos_21": desafios_concluidos_21,
        "desafios_expirados_21": desafios_expirados_21,
        "ultimo_dia_completo": ultimo_dia_completo[:10] if ultimo_dia_completo else None,
        "dias_desde_ultima_atividade": dias_desde,
        "media_xp_dia_21": round(sum_xp_21 / 21, 3),
    }


def features_to_vector(features: dict[str, Any], as_of_date: str) -> dict[str, float]:
    tomorrow = add_days(as_of_date, 1)
    tomorrow_wd = parse_day(tomorrow).weekday()
    tomorrow_wd_js = (tomorrow_wd + 1) % 7
    rates = features.get("weekday_rates") or {}
    row: dict[str, float] = {
        "dias_ativos_7": float(features.get("dias_ativos_7") or 0),
        "dias_ativos_21": float(features.get("dias_ativos_21") or 0),
        "dias_sem_habito": float(features.get("dias_sem_habito") or 0),
        "media_habitos_dia_7": float(features.get("media_habitos_dia_7") or 0),
        "media_habitos_dia_21": float(features.get("media_habitos_dia_21") or 0),
        "taxa_conclusao_7": float(features.get("taxa_conclusao_7") or 0),
        "taxa_conclusao_21": float(features.get("taxa_conclusao_21") or 0),
        "streak_atual": float(features.get("streak_atual") or 0),
        "streak_maximo": float(features.get("streak_maximo") or 0),
        "xp_total": float(features.get("xp_total") or 0),
        "nivel": float(features.get("nivel") or 1),
        "desafios_ativos": float(features.get("desafios_ativos") or 0),
        "desafios_concluidos_21": float(features.get("desafios_concluidos_21") or 0),
        "desafios_expirados_21": float(features.get("desafios_expirados_21") or 0),
        "dias_desde_ultima_atividade": float(
            features.get("dias_desde_ultima_atividade")
            if features.get("dias_desde_ultima_atividade") is not None
            else features.get("dias_sem_habito") or 0
        ),
        "media_xp_dia_21": float(features.get("media_xp_dia_21") or 0),
        "tomorrow_weekday": float(tomorrow_wd_js),
        "tomorrow_rate": float(rates.get(str(tomorrow_wd_js), rates.get(tomorrow_wd_js, 1)) or 1),
    }
    for wd in range(7):
        row[f"wd_{wd}"] = float(rates.get(str(wd), rates.get(wd, 0)) or 0)
    return row


def activity_set(completions: list[dict[str, Any]]) -> set[str]:
    return {c["dia"][:10] for c in completions if c.get("dia")}


def label_y_streak(
    *,
    as_of_date: str,
    streak_atual: int,
    activity_days: set[str],
) -> int | None:
    """1 se streak ativo e gap em D+1..D+3. None se horizonte incompleto."""
    if streak_atual <= 0:
        return 0
    horizon = [add_days(as_of_date, i) for i in range(1, 4)]
    # caller must ensure horizon days exist in dataset window
    for d in horizon:
        if d not in activity_days:
            return 1
    return 0


def label_y_abandono(*, as_of_date: str, activity_days: set[str]) -> int:
    """1 se ≥3 dias consecutivos sem atividade em D+1..D+7."""
    run = 0
    for i in range(1, 8):
        d = add_days(as_of_date, i)
        if d not in activity_days:
            run += 1
            if run >= 3:
                return 1
        else:
            run = 0
    return 0


def estimate_streak_as_of(activity_days: set[str], as_of: str) -> int:
    """Streak aproximado: dias consecutivos com atividade terminando em as_of ou ontem."""
    if as_of in activity_days:
        end = as_of
    elif add_days(as_of, -1) in activity_days:
        end = add_days(as_of, -1)
    else:
        return 0
    streak = 0
    d = end
    while d in activity_days:
        streak += 1
        d = add_days(d, -1)
    return streak


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()

File: src/lib/ml/features.ts
/**
 * ML Fase 1 — feature engineering puro (testável, sem I/O).
 * features_version = "v1"
 */

import { calcularNivel } from "../journey";
import { hojeISO } from "@/lib/datetime";

export const FEATURES_VERSION = "v1";
export const MODEL_VERSION = "heuristic_v1";

export type HabitCompletionRow = { habit_id: string; dia: string; xp_ganho?: number | null };
export type ChallengeAggRow = {
  status: string;
  completed_at?: string | null;
  ends_at?: string | null;
  created_at?: string | null;
};

export type FeatureInput = {
  asOfDate: string; // YYYY-MM-DD (calendário America/Sao_Paulo)
  habitCountAtivo: number;
  completions: HabitCompletionRow[];
  challenges: ChallengeAggRow[];
  streak_atual: number;
  streak_maximo: number;
  xp_total: number;
  ultimo_dia_completo: string | null;
};

export type UserFeaturesV1 = {
  features_version: typeof FEATURES_VERSION;
  dias_ativos_7: number;
  dias_ativos_21: number;
  dias_sem_habito: number;
  media_habitos_dia_7: number;
  media_habitos_dia_21: number;
  taxa_conclusao_7: number;
  taxa_conclusao_21: number;
  weekday_rates: Record<string, number>; // "0".."6" → 0..1
  streak_atual: number;
  streak_maximo: number;
  xp_total: number;
  nivel: number;
  desafios_ativos: number;
  desafios_concluidos_21: number;
  desafios_expirados_21: number;
  ultimo_dia_completo: string | null;
  dias_desde_ultima_atividade: number | null;
  media_xp_dia_21: number;
};

export type MlScoresV1 = {
  model_version: typeof MODEL_VERSION;
  risco_streak: number;
  risco_abandono: number;
  projecao_dias_proximo_nivel: number | null;
  weekday_weakest: number | null;
  explicacao: {
    fatores_streak: string[];
    fatores_abandono: string[];
    weekday_weakest_label: string | null;
  };
};

const WEEKDAY_LABELS = [
  "domingo",
  "segunda",
  "terça",
  "quarta",
  "quinta",
  "sexta",
  "sábado",
] as const;

function parseDay(iso: string): Date {
  return new Date(`${iso}T12:00:00.000Z`);
}

function dayDiff(aIso: string, bIso: string): number {
  const a = parseDay(aIso).getTime();
  const b = parseDay(bIso).getTime();
  return Math.floor((b - a) / (1000 * 60 * 60 * 24));
}

function addDays(iso: string, n: number): string {
  const d = parseDay(iso);
  d.setUTCDate(d.getUTCDate() + n);
  return d.toISOString().slice(0, 10);
}

function clamp01(n: number): number {
  return Math.min(1, Math.max(0, n));
}

function round4(n: number): number {
  return Math.round(n * 10000) / 10000;
}

function round3(n: number): number {
  return Math.round(n * 1000) / 1000;
}

/** Contagens por dia e por weekday a partir de completions. */
export function computeUserFeatures(input: FeatureInput): UserFeaturesV1 {
  const habitN = Math.max(0, input.habitCountAtivo);
  const from7 = addDays(input.asOfDate, -6);
  const from21 = addDays(input.asOfDate, -20);

  const byDay = new Map<string, number>();
  const xpByDay = new Map<string, number>();

  for (const c of input.completions) {
    if (c.dia < from21 || c.dia > input.asOfDate) continue;
    byDay.set(c.dia, (byDay.get(c.dia) ?? 0) + 1);
    const xp = typeof c.xp_ganho === "number" ? c.xp_ganho : 0;
    xpByDay.set(c.dia, (xpByDay.get(c.dia) ?? 0) + xp);
  }

  let diasAtivos7 = 0;
  let diasAtivos21 = 0;
  let sumHabits7 = 0;
  let sumHabits21 = 0;
  let sumXp21 = 0;
  let diasCompletos7 = 0;
  let diasCompletos21 = 0;

  for (let i = 0; i < 21; i++) {
    const dia = addDays(from21, i);
    const count = byDay.get(dia) ?? 0;
    const in7 = dia >= from7;
    if (count > 0) {
      diasAtivos21 += 1;
      if (in7) diasAtivos7 += 1;
    }
    sumHabits21 += count;
    sumXp21 += xpByDay.get(dia) ?? 0;
    if (in7) sumHabits7 += count;
    if (habitN > 0 && count >= habitN) {
      diasCompletos21 += 1;
      if (in7) diasCompletos7 += 1;
    }
  }

  // Weekday rates: for each weekday 0-6, among days in last 21 that fall on that weekday,
  // fraction where at least 1 habit done (or full completion if habits exist)
  const weekdayDone = Array.from({ length: 7 }, () => 0);
  const weekdayTotal = Array.from({ length: 7 }, () => 0);
  for (let i = 0; i < 21; i++) {
    const dia = addDays(from21, i);
    const wd = parseDay(dia).getUTCDay();
    weekdayTotal[wd] += 1;
    const count = byDay.get(dia) ?? 0;
    const ok = habitN > 0 ? count >= Math.max(1, Math.ceil(habitN * 0.5)) : count > 0;
    if (ok) weekdayDone[wd] += 1;
  }
  const weekday_rates: Record<string, number> = {};
  for (let wd = 0; wd < 7; wd++) {
    const t = weekdayTotal[wd] || 1;
    weekday_rates[String(wd)] = round4(weekdayDone[wd] / t);
  }

  let lastActivity: string | null = null;
  for (const dia of [...byDay.keys()].sort()) {
    if ((byDay.get(dia) ?? 0) > 0) lastActivity = dia;
  }
  if (input.ultimo_dia_completo) {
    if (!lastActivity || input.ultimo_dia_completo > lastActivity) {
      lastActivity = input.ultimo_dia_completo;
    }
  }

  const dias_desde_ultima_atividade =
    lastActivity != null ? Math.max(0, dayDiff(lastActivity, input.asOfDate)) : null;

  // Dias sem hábito: consecutive days ending at asOfDate with zero completions
  let dias_sem_habito = 0;
  for (let i = 0; i < 21; i++) {
    const dia = addDays(input.asOfDate, -i);
    if ((byDay.get(dia) ?? 0) === 0) dias_sem_habito += 1;
    else break;
  }

  const from21Chal = `${from21}T00:00:00.000Z`;
  let desafios_ativos = 0;
  let desafios_concluidos_21 = 0;
  let desafios_expirados_21 = 0;
  for (const ch of input.challenges) {
    if (ch.status === "ativo") desafios_ativos += 1;
    if (ch.status === "concluido" && ch.completed_at && ch.completed_at >= from21Chal) {
      desafios_concluidos_21 += 1;
    }
    if (ch.status === "expirado") {
      const ref = ch.ends_at ?? ch.created_at;
      if (ref && ref >= from21Chal) desafios_expirados_21 += 1;
    }
  }

  const level = calcularNivel(input.xp_total);

  return {
    features_version: FEATURES_VERSION,
    dias_ativos_7: diasAtivos7,
    dias_ativos_21: diasAtivos21,
    dias_sem_habito,
    media_habitos_dia_7: round3(sumHabits7 / 7),
    media_habitos_dia_21: round3(sumHabits21 / 21),
    taxa_conclusao_7: habitN > 0 ? round4(diasCompletos7 / 7) : round4(diasAtivos7 / 7),
    taxa_conclusao_21: habitN > 0 ? round4(diasCompletos21 / 21) : round4(diasAtivos21 / 21),
    weekday_rates,
    streak_atual: input.streak_atual,
    streak_maximo: input.streak_maximo,
    xp_total: input.xp_total,
    nivel: level.atual.nivel,
    desafios_ativos,
    desafios_concluidos_21,
    desafios_expirados_21,
    ultimo_dia_completo: input.ultimo_dia_completo,
    dias_desde_ultima_atividade,
    media_xp_dia_21: round3(sumXp21 / 21),
  };
}

export function findWeakestWeekday(rates: Record<string, number>): number | null {
  let weak: number | null = null;
  let min = Infinity;
  for (let wd = 0; wd < 7; wd++) {
    const r = rates[String(wd)];
    if (typeof r !== "number") continue;
    if (r < min) {
      min = r;
      weak = wd;
    }
  }
  return weak;
}

export function scoreUserHeuristicV1(
  features: UserFeaturesV1,
  opts?: { asOfDate?: string; tomorrowWeekday?: number },
): MlScoresV1 {
  const asOf = opts?.asOfDate ?? hojeISO();
  const tomorrowWd =
    opts?.tomorrowWeekday ??
    (() => {
      const d = parseDay(asOf);
      d.setUTCDate(d.getUTCDate() + 1);
      return d.getUTCDay();
    })();

  const weakest = findWeakestWeekday(features.weekday_rates);
  const weakRate = weakest != null ? (features.weekday_rates[String(weakest)] ?? 1) : 1;
  const tomorrowRate = features.weekday_rates[String(tomorrowWd)] ?? 1;

  const fatores_streak: string[] = [];
  let risco_streak = 0;

  if (features.streak_atual <= 0) {
    risco_streak += 0.15;
    fatores_streak.push("sem streak ativo");
  }
  if (features.dias_sem_habito >= 1) {
    risco_streak += Math.min(0.35, 0.12 * features.dias_sem_habito);
    fatores_streak.push(`${features.dias_sem_habito} dia(s) sem hábito`);
  }
  if (features.taxa_conclusao_7 < 0.5) {
    risco_streak += 0.25;
    fatores_streak.push(`taxa 7d baixa (${Math.round(features.taxa_conclusao_7 * 100)}%)`);
  } else if (features.taxa_conclusao_7 < 0.7) {
    risco_streak += 0.12;
    fatores_streak.push(`taxa 7d moderada (${Math.round(features.taxa_conclusao_7 * 100)}%)`);
  }
  if (tomorrowRate < 0.55) {
    risco_streak += 0.22;
    fatores_streak.push(
      `amanhã é ${WEEKDAY_LABELS[tomorrowWd]} (taxa histórica ${Math.round(tomorrowRate * 100)}%)`,
    );
  }
  if (weakest != null && weakRate < 0.45 && tomorrowWd === weakest) {
    risco_streak += 0.1;
    fatores_streak.push(`amanhã coincide com o weekday mais fraco (${WEEKDAY_LABELS[weakest]})`);
  }

  risco_streak = round4(clamp01(risco_streak));

  const fatores_abandono: string[] = [];
  let risco_abandono = 0;
  const idle = features.dias_desde_ultima_atividade ?? features.dias_sem_habito;
  if (idle >= 3) {
    risco_abandono += Math.min(0.45, 0.1 * idle);
    fatores_abandono.push(`${idle} dia(s) sem atividade`);
  }
  if (features.taxa_conclusao_21 > 0.15 && features.taxa_conclusao_7 < features.taxa_conclusao_21 - 0.15) {
    risco_abandono += 0.2;
    fatores_abandono.push("queda de ritmo (7d vs 21d)");
  }
  if (features.desafios_expirados_21 >= 1) {
    risco_abandono += Math.min(0.2, 0.08 * features.desafios_expirados_21);
    fatores_abandono.push(`${features.desafios_expirados_21} desafio(s) expirado(s) em 21d`);
  }
  if (features.dias_ativos_21 <= 3) {
    risco_abandono += 0.15;
    fatores_abandono.push("poucos dias ativos nas últimas 3 semanas");
  }
  risco_abandono = round4(clamp01(risco_abandono));

  const level = calcularNivel(features.xp_total);
  let projecao: number | null = null;
  if (level.proximo && features.media_xp_dia_21 > 0) {
    projecao = Math.max(1, Math.ceil(level.xp_para_proximo / features.media_xp_dia_21));
  }

  return {
    model_version: MODEL_VERSION,
    risco_streak,
    risco_abandono,
    projecao_dias_proximo_nivel: projecao,
    weekday_weakest: weakest,
    explicacao: {
      fatores_streak,
      fatores_abandono,
      weekday_weakest_label: weakest != null ? WEEKDAY_LABELS[weakest] : null,
    },
  };
}

export function formatMlSignalsForMentor(scores: MlScoresV1 | null | undefined): string {
  if (!scores) {
    return "SINAIS ML: ainda não calculados. Não invente previsões numéricas.";
  }
  const riscoStreakPct = Math.round(scores.risco_streak * 100);
  const riscoAbdPct = Math.round(scores.risco_abandono * 100);
  const weak = scores.explicacao.weekday_weakest_label;
  const proj =
    scores.projecao_dias_proximo_nivel != null
      ? `projeção próximo nível ~${scores.projecao_dias_proximo_nivel} dia(s) no ritmo atual`
      : "projeção de nível indisponível (sem XP recente)";
  const fatores = [
    ...scores.explicacao.fatores_streak.slice(0, 3),
    ...scores.explicacao.fatores_abandono.slice(0, 2),
  ];
  return [
    `SINAIS ML (model=${scores.model_version}):`,
    `risco_streak=${riscoStreakPct}%; risco_abandono=${riscoAbdPct}%`,
    weak ? `weekday mais fraco: ${weak}` : null,
    proj,
    fatores.length ? `fatores: ${fatores.join(" | ")}` : null,
    riscoStreakPct >= 55 || riscoAbdPct >= 55
      ? "AÇÃO: priorize presença proativa e concreta — não pergunte só 'como está'."
      : null,
  ]
    .filter(Boolean)
    .join(" ");
}

export { WEEKDAY_LABELS };
